go-controller/pkg/util/pod_annotation.go

```diff
@@ -371,3 +380,326 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
+// Implementations shall return a non-nil pod unless no update is warranted.
+// Implementations can also return a rollback function that will be invoked if
+// the pod update fails.
+type AllocateToPodWithRollbackFunc func(pod *v1.Pod) (*v1.Pod, func(), error)
+
+// UpdatePodAllocationWithRetryOrRollback updates the pod with the result of the
+// allocate function. If the update fails, it applies the rollback provided by
+// the allocate function.
+func UpdatePodAllocationWithRetryOrRollback(podLister listers.PodLister, kube kube.Interface, pod *v1.Pod, allocate AllocateToPodWithRollbackFunc) error {
+	start := time.Now()
+	defer func() {
+		klog.Infof("[%s/%s] pod allocation update took %v", pod.Namespace, pod.Name, time.Since(start))
+	}()
+
+	err := retry.RetryOnConflict(OvnConflictBackoff, func() error {
+		pod, err := podLister.Pods(pod.Namespace).Get(pod.Name)
+		if err != nil {
+			return err
+		}
+
+		// Informer cache should not be mutated, so copy the object
+		pod = pod.DeepCopy()
+		pod, rollback, err := allocate(pod)
+		if err != nil {
+			return err
+		}
+
+		if pod == nil {
+			return nil
+		}
+
+		err = kube.UpdatePod(pod)
+		if err != nil && rollback != nil {
+			rollback()
+		}
+
+		return err
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to update pod %s/%s: %w", pod.Namespace, pod.Name, err)
+	}
+
+	return nil
+}
+
+// UpdatePodAnnotationWithRetry updates the pod annotation on the pod retrying
+// on conflict
+func UpdatePodAnnotationWithRetry(podLister listers.PodLister, kube kube.Interface, pod *v1.Pod, podAnnotation *PodAnnotation, nadName string) error {
+	updatePodAnnotationNoRollback := func(pod *v1.Pod) (*v1.Pod, func(), error) {
+		var err error
+		pod.Annotations, err = MarshalPodAnnotation(pod.Annotations, podAnnotation, nadName)
+		if err != nil {
+			return nil, nil, err
+		}
+		return pod, nil, nil
+	}
+
+	return UpdatePodAllocationWithRetryOrRollback(
+		podLister,
+		kube,
+		pod,
+		updatePodAnnotationNoRollback,
+	)
+}
+
+// ipAllocator is an IP allocation interface that needs to be implemented to use
+// AllocatePodAnnotationWithRollback
+type ipAllocator interface {
+	AllocateIPs(ips []*net.IPNet) error
+	AllocateNextIPs() ([]*net.IPNet, error)
+	ReleaseIPs(ips []*net.IPNet) error
+	IsErrAllocated(error) bool
+}
+
+// AllocatePodAnnotationWithRollback allocates the pod annotation which includes
+// IPs, a mac address and gateways. The allocation can be requested through the
+// network selection element or derived from the allocator provided IPs. If the
+// requested IPs cannot be honored, a new set of IPs will be allocated unless
+// reallocate is set to false. A rollback function is returned to rollback the
+// IP allocation if there was one. This function is designed to be used in
+// AllocateToPodWithRollbackFunc implementations. Use an inlined implementation
+// if you want to extract information from it as a side-effect.
+func AllocatePodAnnotationWithRollback(
```

Comment:
Don't like `GeneratePodAnnotation` either.

A PodAnnotation is allocated by extension of an IP being allocated and all the contained information being derived from that allocation in a way that two running pods would not have the same PodAnnotation.

The meaning of Allocation here comes with the added connotation that you can `undo` through the rollback and I think it is important to carry that connotation forward without trying to read to much into it otherwise.

It might not be perfect and  I will take further suggestions to try to come up with a better name but I think this is or could be sufficiently explained in the doc of the method.